src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -161,13 +170,123 @@ static unsigned uct_cuda_copy_iface_progress(uct_iface_h tl_iface)
     unsigned max_events = iface->config.max_poll;
     unsigned count;
 
-    count = uct_cuda_copy_progress_event_queue(&iface->outstanding_d2h_cuda_event_q,
+    count = uct_cuda_copy_progress_event_queue(iface, UCT_CUDA_COPY_D2H_STREAM,
                                                max_events);
-    count += uct_cuda_copy_progress_event_queue(&iface->outstanding_h2d_cuda_event_q,
+    count += uct_cuda_copy_progress_event_queue(iface, UCT_CUDA_COPY_H2D_STREAM,
                                                 (max_events - count));
     return count;
 }
 
+static ucs_status_t uct_cuda_copy_iface_event_fd_get(uct_iface_h tl_iface, int *fd_p)
+{
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+
+    if (-1 == iface->eventfd) {
+        iface->eventfd = eventfd(0, EFD_CLOEXEC | EFD_NONBLOCK);
```

Comment:
need to use https://github.com/openucx/ucx/blob/master/src/uct/api/uct.h#L1041, and not create extra fd

Reply:
@yosefe Looks like `async_event_cb` is needed to support UCT_IFACE_FLAG_EVENT_ASYNC_CB but this PR enables `cuda_copy` UCT to support UCT_IFACE_FLAG_EVENT_FD. Do we need `async_event_cb` for this? 

For RC, I see that the fd that it uses to support UCT_IFACE_FLAG_EVENT_FD it uses ibv_comp_channel's fd member. Instead cuda_copy does what sm UCT [does](https://github.com/openucx/ucx/blob/master/src/uct/sm/mm/base/mm_iface.c#L483) by creating a new fd. PLMK if I'm missing something.

Reply:
in SM, the ginalling really happens between two different processes. in cuda_copy, AFAIK eventfd is used to signal between 2 threads of same process: https://github.com/openucx/ucx/pull/5283/files#diff-885841dbead9e3676911f271698fda25R204
so in that place, we should call the async_cb instead

Reply:
Will try to jot down what this [write](https://github.com/openucx/ucx/pull/5283/files#diff-885841dbead9e3676911f271698fda25R204) call does so that we're on the same page. This write is used relay a notification to the agent which sleeps waiting for an event on the fd (likely ucp_worker after arm and epoll_wait). The write happens from an internal thread that results from [cuda stream callback mechanism](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__STREAM.html#group__CUDA__STREAM_1g613d97a277d7640f4cb1c03bd51c2483) which explicitly manages the thread (not UCT) and is triggered when preceding events in the cuda stream completes. In this case, when a `cudaMemcpy` operation on a stream completes, cuda runtime managed thread runs the function that results in the write and subsequent event notification. So in that sense, this write isn't used to signal between 2 threads explicitly managed by `cuda_copy` UCT but at the end of the day it is signalling between 2 threads. Also, `cuda_copy` doesn't explicitly create an asynchronous event handler as such. Do we still need async_cb in this case?

Reply:
yes, i think async_cb will be a simpler solution: instead of creating event_fd, writing to it, etc. just call a callback
async_cb was created for this cases, where the event arrives on some thread, and we need to wake up the main thread.